lib/ukintctlr/include/uk/intctlr.h

```diff
@@ -1,42 +1,187 @@
+	void (*configure_irq)(struct uk_intctlr_irq *irq);
+	void (*handle_irq)(struct __regs *regs, unsigned int irq);
+	void (*mask_irq)(unsigned int irq);
+	void (*unmask_irq)(unsigned int irq);
+	void (*set_irq_trigger)(unsigned int irq, unsigned int trigger);
+};
+
+UK_CTASSERT(__offsetof(struct uk_intctlr_driver_ops, handle_irq) ==
+	    UK_INTCTLR_DRIVER_OPS_HANDLE_IRQ_OFFS);
+
+/** Interrupt controller descriptor */
+struct uk_intctlr_desc {
+	char *name;
+	struct uk_intctlr_driver_ops *ops;
+};
+
+UK_CTASSERT(__offsetof(struct uk_intctlr_desc, ops) == UK_INTCTLR_DESC_OPS_OFFS);
+
+/** Interrupt handler function */
+typedef int (*uk_intctlr_irq_handler_func_t)(void *);
+
+/**
+ * Probe the interrupt controller
+ *
+ * This function must be implemented by the driver
+ *
+ * @return zero on success or negative value on error
+ */
+int uk_intctlr_probe(void);
+
+/**
+ * Handle an interrupt
+ *
+ * This function must be implemented by the driver
+ *
+ * @param regs Register context at the time the interrupt was raised
+ * @param irq  Interrupt to handle
+ * @return zero on success or negative value on error
+ */
+void uk_intctlr_irq_handle(struct __regs *regs, unsigned int irq);
+
+/**
+ * Register interrupt controller driver with the uk_intctlr subsystem
+ *
+ * This function must be called by the driver during probe
+ *
+ * @param intctlr populated interrupt controller descriptor
+ * @return zero on success or negative value on error
+ */
+int uk_intctlr_register(struct uk_intctlr_desc *intctlr);
+
+/**
+ * Initialize the uk_intctlr subsystem
+ *
+ * Must be called after uk_intctlr_probe
+ *
+ * @param alloc The allocator to use for internal allocations
+ * @return zero on success, negative value on failure
+ */
+int uk_intctlr_init(struct uk_alloc *alloc);
+
+/**
+ * Register an interrupt handler
+ *
+ * @param irq     Interrupt to register handler for
+ * @param handler Handler function
+ * @param arg     Caller data to be passed to the handler
+ */
+int uk_intctlr_irq_register(unsigned int irq,
+			    uk_intctlr_irq_handler_func_t handler,
+			    void *arg);
+
+/**
+ * Unregister a previously registered interrupt handler
+ *
+ * @param irq     Interrupt to register handler for
+ * @param handler Handler function
+ */
+int uk_intctlr_irq_unregister(unsigned int irq,
+			      uk_intctlr_irq_handler_func_t handler);
+
+/**
+ * Allocate IRQs from available pool
+ *
+ * @param irqs pointer to array of irqs
+ * @param sz   number of array elements
+ * @return zero on success, or negative value on error
+ */
+int uk_intctlr_irq_alloc(unsigned int *irqs, __sz count);
+
+/**
+ * Free previously allocated IRQs
+ *
+ * @param irqs pointer to array of irqs
+ * @param sz   number of array elements
+ * @return zero on success, or negative value on error
+ */
+int uk_intctlr_irq_free(unsigned int *irqs, __sz count);
+
+/**
+ * Parse `interrupts` fdt node into IRQ descriptor
+ *
+ * This function must be implemented by the driver
```

Comment:
Yes, you're right. See also the discussion in https://github.com/unikraft/unikraft/pull/1103#discussion_r1332840305, but either way it will be only applicable to cases that use fdt.